#!/bin/bash
set -e

# Get the current directory
INSTALLATION_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
# Get the parent directory which should be the main directory of ProteoSeeker.
PS_DIR=$(dirname "${INSTALLATION_DIR}")
PARAM_DIR="${PS_DIR}/parameter_files"
# The name of the parameter file.
PARAMETER_FILE="${PARAM_DIR}/par_seek_c.txt"

# Create the parameter files directory if needed.
if [ ! -d "${PARAM_DIR}" ]; then
    mkdir "${PARAM_DIR}"
fi

# Sourcing
source "${INSTALLATION_DIR}/find_conda.sh"
source "${INSTALLATION_DIR}/find_mb_cb.sh"
source "${INSTALLATION_DIR}/find_pd.sh"

# Seek - Contigs
# Creating the file for the input paths and option and writing in it.
echo "#---------Input and output options---------" > "${PARAMETER_FILE}"
echo "input_folder=\"input\"" >> "${PARAMETER_FILE}"
echo "sra_code=\"SRR12829170\"" >> "${PARAMETER_FILE}"
echo "contigs=\"True\"" >> "${PARAMETER_FILE}"
echo "protein_input=\"\"" >> "${PARAMETER_FILE}"
echo "adapters_path=\"${PS_DIR}/adapters.fa\"" >> "${PARAMETER_FILE}"
echo "protein_db_path=\"${PD_PATH}\"" >> "${PARAMETER_FILE}"
echo "kraken_db_path=\"${PS_DIR}/ps_tools/kraken2/kraken2_databases/kraken2_8st_db\"" >> "${PARAMETER_FILE}"
echo "profiles_path=\"\"" >> "${PARAMETER_FILE}"
echo "profiles_phylo_path=\"\"" >> "${PARAMETER_FILE}"
echo "profiles_broad_path=\"${PS_DIR}/pfam_database/Pfam-A.hmm\"" >> "${PARAMETER_FILE}"
echo "swissprot_path=\"${PS_DIR}/swissprot_database/swissprot\"" >> "${PARAMETER_FILE}"
echo "motifs_path=\"${PS_DIR}/motifs.txt\"" >> "${PARAMETER_FILE}"
echo "output_path=\"${PS_DIR}/results/SRR12829170_seek_c\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------Modes and Routes---------" >> "${PARAMETER_FILE}"
echo "seek_mode=\"True\"" >> "${PARAMETER_FILE}"
echo "taxonomy_mode=\"False\"" >> "${PARAMETER_FILE}"
echo "seek_route=\"1\"" >> "${PARAMETER_FILE}"
echo "taxonomy_route=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------Protein family options---------" >> "${PARAMETER_FILE}"
echo "seek_family_code=\"1191,3957\"" >> "${PARAMETER_FILE}"
echo "taxonomy_family_code=\"\"" >> "${PARAMETER_FILE}"
echo "seek_db_name=\"cas_als\"" >> "${PARAMETER_FILE}"
echo "taxonomy_db_name=\"\"" >> "${PARAMETER_FILE}"
echo "input_seek_protein_names_status=\"\"" >> "${PARAMETER_FILE}"
echo "input_seek_protein_names=\"\"" >> "${PARAMETER_FILE}"
echo "input_taxonomy_protein_names_status=\"\"" >> "${PARAMETER_FILE}"
echo "input_taxonomy_protein_names=\"\"" >> "${PARAMETER_FILE}"
echo "name_threshold=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------Pipeline---------" >> "${PARAMETER_FILE}"
echo "seek_route=\"1\"" >> "${PARAMETER_FILE}"
echo "paired_end=\"False\"" >> "${PARAMETER_FILE}"
echo "compressed=\"False\"" >> "${PARAMETER_FILE}"
echo "prefetch_size=\"\"" >> "${PARAMETER_FILE}"
echo "adapters_status=\"\"" >> "${PARAMETER_FILE}"
echo "add_seek_info=\"\"" >> "${PARAMETER_FILE}"
echo "add_taxonomy_info=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------FastQC---------" >> "${PARAMETER_FILE}"
echo "skip_fastqc=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------BBDuk---------" >> "${PARAMETER_FILE}"
echo "bbduk_max_ram=\"\"" >> "${PARAMETER_FILE}"
echo "clear_space=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------Megahit---------" >> "${PARAMETER_FILE}"
echo "k_list=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------Kraken---------" >> "${PARAMETER_FILE}"
echo "kraken_threshold=\"\"" >> "${PARAMETER_FILE}"
echo "kraken_memory_mapping=\"\"" >> "${PARAMETER_FILE}"
echo "bracken_length=\"\"" >> "${PARAMETER_FILE}"
echo "bracken_level=\"\"" >> "${PARAMETER_FILE}"
echo "bracken_threshold=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------Binning---------" >> "${PARAMETER_FILE}"
echo "binning_tool=\"\"" >> "${PARAMETER_FILE}"
echo "binning_max_ram=\"\"" >> "${PARAMETER_FILE}"
echo "bin_contig_len=\"1000\"" >> "${PARAMETER_FILE}"
echo "bin_kmer=\"\"" >> "${PARAMETER_FILE}"
echo "comebin_batch_size=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------CD-HIT---------" >> "${PARAMETER_FILE}"
echo "cd_hit_t=\"\"" >> "${PARAMETER_FILE}"
echo "cd_hit_max_mem=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------Gene prediction---------" >> "${PARAMETER_FILE}"
echo "gene_encoding=\"\"" >> "${PARAMETER_FILE}"
echo "genetic_code=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------HMMER---------" >> "${PARAMETER_FILE}"
echo "val_type=\"\"" >> "${PARAMETER_FILE}"
echo "second_dom_search=\"\"" >> "${PARAMETER_FILE}"
echo "e_value_nodom_thr=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------Annotation---------" >> "${PARAMETER_FILE}"
echo -e "add_type=\"organism\tbias\tnumber of organisms\tbiosample name\"" >> "${PARAMETER_FILE}"
echo -e "add_info=\"synthetic metagenome\tno\t10\tSAMN16441671\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------Threads---------" >> "${PARAMETER_FILE}"
echo "thread_num=\"4\"" >> "${PARAMETER_FILE}"
echo "filtering_threads=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------Processes performed after---------" >> "${PARAMETER_FILE}"
echo "after_preprocessing=\"\"" >> "${PARAMETER_FILE}"
echo "after_assembly=\"\"" >> "${PARAMETER_FILE}"
echo "after_gene_pred=\"\"" >> "${PARAMETER_FILE}"
echo "after_binning=\"\"" >> "${PARAMETER_FILE}"
echo "after_db=\"\"" >> "${PARAMETER_FILE}"
echo "after_tm=\"\"" >> "${PARAMETER_FILE}"
echo "after_ap=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------Processes performed up to---------" >> "${PARAMETER_FILE}"
echo "up_to_sra=\"\"" >> "${PARAMETER_FILE}"
echo "up_to_databases=\"\"" >> "${PARAMETER_FILE}"
echo "up_to_preprocessing_com=\"\"" >> "${PARAMETER_FILE}"
echo "up_to_preprocessing_uncom=\"\"" >> "${PARAMETER_FILE}"
echo "up_to_alignment=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------Tool enviroments---------" >> "${PARAMETER_FILE}"
echo "sra_env=\"\"" >> "${PARAMETER_FILE}"
echo "fastqc_env=\"\"" >> "${PARAMETER_FILE}"
echo "bbduk_env=\"\"" >> "${PARAMETER_FILE}"
echo "megahit_env=\"\"" >> "${PARAMETER_FILE}"
echo "kraken_env=\"\"" >> "${PARAMETER_FILE}"
echo "bracken_env=\"\"" >> "${PARAMETER_FILE}"
echo "metabinner_env=\"${MTBENV}\"" >> "${PARAMETER_FILE}"
echo "comebin_env=\"\"" >> "${PARAMETER_FILE}"
echo "cdhit_env=\"\"" >> "${PARAMETER_FILE}"
echo "genepred_env=\"\"" >> "${PARAMETER_FILE}"
echo "hmmer_env=\"\"" >> "${PARAMETER_FILE}"
echo "diamond_env=\"\"" >> "${PARAMETER_FILE}"
echo "taxonkit_env=\"\"" >> "${PARAMETER_FILE}"
echo "phobius_env=\"\"" >> "${PARAMETER_FILE}"
echo "bowtie_env=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"

echo "#---------Tool paths---------" >> "${PARAMETER_FILE}"
echo "conda_bin=\"${CONDA_INST_DIR}\"" >> "${PARAMETER_FILE}"
echo "conda_sh=\"${CONDA_SH_PATH}\"" >> "${PARAMETER_FILE}"
echo "prefetch_path=\"\"" >> "${PARAMETER_FILE}"
echo "vdb_validate_path=\"\"" >> "${PARAMETER_FILE}"
echo "fastq_dump_path=\"\"" >> "${PARAMETER_FILE}"
echo "fastqc_path=\"\"" >> "${PARAMETER_FILE}"
echo "gzip_path=\"\"" >> "${PARAMETER_FILE}"
echo "cat_path=\"\"" >> "${PARAMETER_FILE}"
echo "bbduk_path=\"\"" >> "${PARAMETER_FILE}"
echo "megahit_path=\"\"" >> "${PARAMETER_FILE}"
echo "kraken_path=\"\"" >> "${PARAMETER_FILE}"
echo "bracken_path=\"\"" >> "${PARAMETER_FILE}"
echo "alpha_diversity_path=\"ps_tools/KrakenTools/DiversityTools/alpha_diversity.py\"" >> "${PARAMETER_FILE}"
echo "metabinner_bin_path=\"${MBINPATH}\"" >> "${PARAMETER_FILE}"
echo "comebin_bin_path=\"${CBINPATH}\"" >> "${PARAMETER_FILE}"
echo "cd_hit_path=\"\"" >> "${PARAMETER_FILE}"
echo "fraggenescanrs_path=\"${PS_DIR}/ps_tools/fgsrs/FragGeneScanRs\"" >> "${PARAMETER_FILE}"
echo "hmmscan_path=\""\" >> "${PARAMETER_FILE}"
echo "hmmpress_path=\"\"" >> "${PARAMETER_FILE}"
echo "hmmfetch_path=\"\"" >> "${PARAMETER_FILE}"
echo "diamond_path=\"\"" >> "${PARAMETER_FILE}"
echo "taxonkit_path=\"\"" >> "${PARAMETER_FILE}"
echo "phobius_path=\"\"" >> "${PARAMETER_FILE}"
echo "bowtie_build_path=\"\"" >> "${PARAMETER_FILE}"
echo "bowtie_path=\"\"" >> "${PARAMETER_FILE}"
echo "" >> "${PARAMETER_FILE}"